### ISSUE TITLE
Cannot read properties of null (reading 'chainId')

### ISSUE THREAD

**Issue by adrienthiery** [OpenZeppelin/openzeppelin-upgrades#943]
I'm getting this error out of the blue with the latest versions when running: `npx hardhat run --network private scripts/deploy.js`
 
```TypeError: Cannot read properties of null (reading 'chainId')
    at new Manifest (/Users/adrien/Code/my-project/my-project-smart-contracts/node_modules/@openzeppelin/upgrades-core/src/manifest.ts:133:65)
    at Function.forNetwork (/Users/adrien/Code/my-project/my-project-smart-contracts/node_modules/@openzeppelin/upgrades-core/src/manifest.ts:104:14)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at Proxy.deployProxy (/Users/adrien/Code/my-project/my-project-smart-contracts/node_modules/@openzeppelin/hardhat-upgrades/src/deploy-proxy.ts:46:22)
    at deployMyProjectNFT (/Users/adrien/Code/my-project/my-project-smart-contracts/scripts/deploy.js:106:21)
    at main (/Users/adrien/Code/my-project/my-project-smart-contracts/scripts/deploy.js:29:5)
    at /Users/adrien/Code/my-project/my-project-smart-contracts/scripts/deploy.js:128:7
```

When logging this is the node modules, I find this is on [this line](https://github.com/OpenZeppelin/openzeppelin-upgrades/blob/6a89da4ee700aa9dcae96e669d93b7f8c38b8728/packages/core/src/manifest.ts#L133) and if I log `devInstanceMetadata`, I get: 
```
{
  networkName: 'hardhat',
  instanceId: '0xb7a74718b581ff4054338676a6cda012eb69bdccaab5f739f66eb1a8f6e5362e',
  forkedNetwork: null
}
```

<details>
  <summary>My `hardhat.config.js` is as follows</summary> 
  
```
require("hardhat-contract-sizer");
require("@nomicfoundation/hardhat-toolbox");
require("@openzeppelin/hardhat-upgrades");
require("hardhat-gas-reporter");
require("hardhat-tracer");

const path = require("path");

let BUILD_DIR = path.join(__dirname, "../my-project-frontend/public/hardhat/");
if (process.env.BUILD_DIR) {
  BUILD_DIR = process.env.BUILD_DIR;
}

let COIN_MARKET_CAP_API_KEY = "some-key";
if (process.env.COIN_MARKET_CAP_API_KEY) {
  COIN_MARKET_CAP_API_KEY = process.env.COIN_MARKET_CAP_API_KEY;
}

const PRIVATE_KEY = process.env.PRIVATE_KEY;
const OPTIMIZATION_RUNS = 200;

/** @type import('hardhat/config').HardhatUserConfig */
module.exports = {
  paths: {
    artifacts: BUILD_DIR,
  },

  etherscan: {
    apiKey: {
      private: process.env.ETHERSCAN_API_KEY,
      localhost: "whatever",
    },
    customChains: [
      {
        network: "localhost",
        name: "localhost",
        chainId: 1337,
        urls: {
          apiURL: "http://localhost:4000/api",
          browserURL: "http://localhost:4000/",
        },
      },
      {
        network: "private",
        name: "private",
        chainId: 8888,
        urls: {
          apiURL: "https://<private block explorer>/api",
          browserURL: "https://<private block explorer>/",
        },
      },
    ],
  },

  networks: {
    hardhat: {
      chainId: 1337,
      ...(process.env.NODE_ENV !== "test"
        ? {
            mining: {
              auto: false,
              interval: 1000,
            },
          }
        : {}),
    },

    localhost: {
      timeout: 500_000, // required when running some upgrade tests
    },

    private: {
      url: "https://<private-rpc>",
      accounts: PRIVATE_KEY ? [PRIVATE_KEY] : [],
    },
  },

  solidity: {
    version: "0.8.20",
    settings: {
      optimizer: {
        enabled: true,
        runs: OPTIMIZATION_RUNS,
        details: {
          yul: false,
        },
      },
    },
  },

  contractSizer: {
    alphaSort: true,
    disambiguatePaths: false,
    runOnCompile: true,
  },

  gasReporter: {
    enabled: true,
    coinmarketcap: COIN_MARKET_CAP_API_KEY,
    outputFile: "gas-report.out",
    forceConsoleOutput: true,
  },
};
```

</details> 

<details>
  <summary>My `deploy.js` script</summary> 
  
```
// We require the Hardhat Runtime Environment explicitly here. This is optional
// but useful for running the script in a standalone fashion through `node <script>`.
//
// You can also run a script with `npx hardhat run <script>`. If you do that, Hardhat
// will compile your contracts, add the Hardhat Runtime Environment's members to the
// global scope, and execute the script.
const hre = require("hardhat");
const {
  enhanceArtifactsAndVerifyContracts,
  myProjectDeploymentLogger,
  upgradeOptions,
  sendIrrelevantTransactions,
} = require("./utils/deployment-utils");
const { ethers, upgrades } = require("hardhat");

async function main() {
  const deployer = (await hre.ethers.getSigners())[0];

  // DEVELOPMENT ONLY
  // Do a random number of irrelevant transactions with the deployer
  // so that we increase transaction count and make the contract address random
  // upon redeployment
  if (hre.network.name === "localhost") {
    const randomNumber = Math.floor(Math.random() * 21);
    await sendIrrelevantTransactions(deployer, randomNumber);
  }

  const { MyProjectNFT, implementationAddress: MyProjectNFTImplementationAddress } =
    await deployMyProjectNFT(deployer);
  const { MyProject, implementationAddress: MyProjectImplementationAddress } =
    await deployMyProject(deployer, MyProjectNFT);

  await enhanceArtifactsAndVerifyContracts([
    {
      name: "MyProjectNFT",
      contract: MyProjectNFT,
      args: [],
      implementationAddress: MyProjectNFTImplementationAddress,
    },
    {
      name: "MyProject",
      contract: MyProject,
      args: [],
      implementationAddress: MyProjectImplementationAddress,
    },
  ]);

  myProjectDeploymentLogger("[+] DEPLOYMENT OF ALL CONTRACTS SUCCESSFUL");

  return {
    MyProjectNFT,
    MyProject,
  };
}

async function deployMyProject(deployer, MyProjectNFT) {
  const paymentDelayDuration = process.env.PAYMENT_DELAY_DURATION || 60; // 1 minute
  const cooldownPeriod = process.env.COOLDOWN_PERIOD || 10; // 10 seconds
  const epochCost = process.env.EPOCH_COST || 100; // 100 WEI
  const epochTime = process.env.EPOCH_TIME || 60; // 1 minute
  const burningRateStep =
    process.env.BURNING_RATE_STEP || hre.ethers.parseEther("0.0001"); // 0.0001 ETH
  const protocolFeePercentage = process.env.PROTOCOL_FEE_PERCENTAGE || 5;
  const protocolFeeDestination =
    process.env.PROTOCOL_FEE_DESTINATION || deployer?.address;

  const myProjectNFTProxyAddress = await MyProjectNFT.getAddress();

  const myProjectArgs = [
    deployer.address,
    myProjectNFTProxyAddress,
    paymentDelayDuration,
    cooldownPeriod,
    epochCost,
    epochTime,
    protocolFeePercentage,
    burningRateStep,
    protocolFeeDestination,
  ];

  const MyProjectFactory = await ethers.getContractFactory("MyProject");
  const MyProject = await upgrades.deployProxy(
    MyProjectFactory,
    myProjectArgs,
    upgradeOptions
  );
  await MyProject.waitForDeployment();

  myProjectDeploymentLogger("-------------------------------------------");
  const myProjectProxyAddress = await MyProject.getAddress();
  const implementationAddress = await upgrades.erc1967.getImplementationAddress(
    myProjectProxyAddress
  );
  myProjectDeploymentLogger(
    `MyProject contract deployed at ${myProjectProxyAddress} (impl: ${implementationAddress})`
  );
  myProjectDeploymentLogger("-------------------------------------------");

  await MyProjectNFT.setTriggerContract(myProjectProxyAddress);

  return { MyProject, implementationAddress };
}

async function deployMyProjectNFT(deployer) {
  const MyProjectNFTFactory = await hre.ethers.getContractFactory("MyProjectNFT");
  const MyProjectNFT = await upgrades.deployProxy(
    MyProjectNFTFactory,
    [deployer.address],
    upgradeOptions
  );
  await MyProjectNFT.waitForDeployment();

  myProjectDeploymentLogger("-------------------------------------------");
  const myProjectNFTProxyAddress = await MyProjectNFT.getAddress();
  const implementationAddress = await upgrades.erc1967.getImplementationAddress(
    myProjectNFTProxyAddress
  );
  myProjectDeploymentLogger(
    `MyProjectNFT contract deployed at ${myProjectNFTProxyAddress} (impl: ${implementationAddress})`
  );
  myProjectDeploymentLogger("-------------------------------------------");
  return { MyProjectNFT, implementationAddress };
}

if (require.main === module) {
  (async () => {
    try {
      await main();
    } catch (e) {
      console.error(e);
      process.exitCode = 1;
    } finally {
      process.exit();
    }
  })();
}

module.exports = {
  deploy: main,
};
```

</details> 

Any idea what's going wrong? Or if that comes from my config? 
Why would the metadata give something about the `hardhat` network? 🤔 

Thanks a lot for the help!

**Comment by ericglau** [comment]
- Are you using a forked network?
- Can you share what versions of `hardhat` and `@openzeppelin/hardhat-upgrades` and `@openzeppelin/upgrades-core` packages you are using?  e.g. post the output of `npm list`
- Can you post the output of this command, when your Hardhat node is running?
```
curl -X POST --data '{"jsonrpc":"2.0","method":"hardhat_metadata","params":[],"id":1}' localhost:8545
```

**Comment by adrienthiery** [comment]
Hi @ericglau,

> * Are you using a forked network?

Nope, this is an `anvil` chain started without forking

> * Can you share what versions of `hardhat` and `@openzeppelin/hardhat-upgrades` and `@openzeppelin/upgrades-core` packages you are using?  e.g. post the output of `npm list`

```
├── @nomicfoundation/hardhat-chai-matchers@2.0.2
├── @nomicfoundation/hardhat-ethers@3.0.5
├── @nomicfoundation/hardhat-network-helpers@1.0.9
├── @nomicfoundation/hardhat-toolbox@3.0.0
├── @nomicfoundation/hardhat-verify@1.1.1
├── @nomiclabs/hardhat-etherscan@3.1.7
├── @openzeppelin/contracts-upgradeable@5.0.0
├── @openzeppelin/hardhat-upgrades@2.4.1
```

> * Can you post the output of this command, when your Hardhat node is running?
> ```
> curl -X POST --data '{"jsonrpc":"2.0","method":"hardhat_metadata","params":[],"id":1}' localhost:8545
> ```

Here is the output:
```
{"jsonrpc":"2.0","id":null,"error":{"code":-32600,"message":"Invalid request"}}
```

**Comment by adrienthiery** [comment]
In the meantime, we patched `node_modules/@openzeppelin/upgrades-core/dist/manifest.js` to change it from `if (devInstanceMetadata.forkedNetwork !== undefined)` to  `if (devInstanceMetadata.forkedNetwork)` with:
```sed -i "s/devInstanceMetadata.forkedNetwork !== undefined/devInstanceMetadata.forkedNetwork/g" node_modules/@openzeppelin/upgrades-core/dist/manifest.js```

Is there any reason to differentiate between `undefined` and `null` in there @ericglau ? 

Seems to do the trick although not ideal

**Comment by ericglau** [comment]
Thanks for sharing this information.

We call the [RPC method `hardhat_metadata`](https://hardhat.org/hardhat-network/docs/reference#hardhat_metadata) to check if it is a Hardhat node.  We normally only expected Hardhat to have this method, while other networks should give an error.

So in the Upgrades plugin, the [getDevInstanceMetadata](https://github.com/OpenZeppelin/openzeppelin-upgrades/blob/11e37ebf7b97731425b4bc6c93c7792ada540c8c/packages/core/src/manifest.ts#L47) function should return [undefined](https://github.com/OpenZeppelin/openzeppelin-upgrades/blob/11e37ebf7b97731425b4bc6c93c7792ada540c8c/packages/core/src/manifest.ts#L56).  So we would not expect the possibility for it to reach a case where `devInstanceMetadata.forkedNetwork` is null.  Hardhat would normally only have that defined or undefined.

Are your Anvil nodes configured in any particular way?

@adrienthiery Could you try creating a Hardhat script with the following contents, and run it while targeting your Anvil node to see what is the output?
```
import { ethers } from "hardhat";

async function main() {
  const result = await ethers.provider.send('hardhat_metadata', []);
  console.log('result', JSON.stringify(result, null, 2));
}

// We recommend this pattern to be able to use async/await everywhere
// and properly handle errors.
main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
```

**Comment by adrienthiery** [comment]
> Are your Anvil nodes configured in any particular way?

No special configuration except we enabled auto-mining.

Sure, here it is: 
```
result {
  "clientVersion": "anvil/v0.2.0",
  "chainId": 8888,
  "instanceId": "0xb7a74718b581ff4054338676a6cda012eb69bdccaab5f739f66eb1a8f6e5362e",
  "latestBlockNumber": 57614,
  "latestBlockHash": "0xa848d54d7edc30bb75ebbc1c8d5bf8934edcd49cc33c7432d3408528dbbaf13a",
  "forkedNetwork": null,
  "snapshots": {}
}
```

**Comment by ericglau** [comment]
This is fixed in `@openzeppelin/upgrades-core@1.32.2`

**Comment by adrienthiery** [comment]
Awesome, thanks!
